tests/v2/validation/provisioning/rke2/cert_rotation_test.go

```diff
@@ -0,0 +1,173 @@
+	"github.com/stretchr/testify/require"
+	"github.com/stretchr/testify/suite"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/apimachinery/pkg/watch"
+)
+
+type V2ProvCertRotationTestSuite struct {
+	suite.Suite
+	session     *session.Session
+	client      *rancher.Client
+	config      *provisioning.Config
+	clusterName string
+	namespace   string
+}
+
+func (r *V2ProvCertRotationTestSuite) TearDownSuite() {
+	r.session.Cleanup()
+}
+
+func (r *V2ProvCertRotationTestSuite) SetupSuite() {
+	testSession := session.NewSession(r.T())
+	r.session = testSession
+
+	r.config = new(provisioning.Config)
+	config.LoadConfig(provisioning.ConfigurationFileKey, r.config)
+
+	client, err := rancher.NewClient("", testSession)
+	require.NoError(r.T(), err)
+
+	r.client = client
+
+	r.clusterName = r.client.RancherConfig.ClusterName
+	r.namespace = r.client.RancherConfig.ClusterName
+}
+
+func (r *V2ProvCertRotationTestSuite) testCertRotation(provider Provider, kubeVersion string, nodesAndRoles []machinepools.NodeRoles, credential *cloudcredentials.CloudCredential) {
+	name := fmt.Sprintf("Provider_%s/Kubernetes_Version_%s/Nodes_%v", provider.Name, kubeVersion, nodesAndRoles)
+	r.Run(name, func() {
+		r.Run("initial", func() {
+			testSession := session.NewSession(r.T())
+			defer testSession.Cleanup()
+
+			testSessionClient, err := r.client.WithSession(testSession)
+			require.NoError(r.T(), err)
+
+			clusterName := provisioning.AppendRandomString(fmt.Sprintf("%s-%s", r.clusterName, provider.Name))
+			generatedPoolName := fmt.Sprintf("nc-%s-pool1-", clusterName)
+			machinePoolConfig := provider.MachinePoolFunc(generatedPoolName, namespace)
+
+			machineConfigResp, err := machinepools.CreateMachineConfig(provider.MachineConfig, machinePoolConfig, testSessionClient)
+			require.NoError(r.T(), err)
+
+			machinePools := machinepools.RKEMachinePoolSetup(nodesAndRoles, machineConfigResp)
+
+			cluster := clusters.NewRKE2ClusterConfig(clusterName, namespace, "calico", credential.ID, kubeVersion, machinePools)
+			clusterResp, err := clusters.CreateRKE2Cluster(testSessionClient, cluster)
+			require.NoError(r.T(), err)
+
+			kubeProvisioningClient, err := r.client.GetKubeAPIProvisioningClient()
+			require.NoError(r.T(), err)
+
+			result, err := kubeProvisioningClient.Clusters("fleet-default").Watch(context.TODO(), metav1.ListOptions{
+				FieldSelector:  "metadata.name=" + cluster.ObjectMeta.Name,
+				TimeoutSeconds: &defaults.WatchTimeoutSeconds,
+			})
+			require.NoError(r.T(), err)
+			checkFunc := clusters.IsProvisioningClusterReady
+
+			err = wait.WatchWait(result, checkFunc)
+			require.NoError(r.T(), err)
+			assert.Equal(r.T(), clusterName, clusterResp.ObjectMeta.Name)
+
+			cluster, err = r.client.Provisioning.Cluster.ByID(clusterResp.ID)
+			require.NoError(r.T(), err)
+			require.NotNil(r.T(), cluster.Status)
+
+			// rotate certs
+			require.NoError(r.T(), r.rotateCerts(clusterName, 1))
+			// verify certs rotated
+
```

Comment:
The comment `// verify certs rotated` makes me think there is some logic missing here, is there? I can see that `r.rotateCerts` waits for the `CertificateRotationGeneration` to equal the expected generation, is that enough to verify they have been rotated? Same question in regards to the below comment `// wait until cluster Updated == True`.

Reply:
I was going to add some logic to scrape the certs from the node to verify that they were indeed rotated, but realistically we don't need to do that, verifying the generation is all we need since that would indicate a bug in rke2/k3s, which I think is out of the scope of these tests since a failure to rotate should cause the cert rotation operation to fail/not complete. I've removed the comments.